.github/workflows/ci.yml

```diff
@@ -0,0 +1,87 @@
+name: ci
+
+on:
+  pull_request:
+  push:
+
+jobs:
+  check:
+    runs-on: ubuntu-latest
+    outputs:
+      branch: ${{ steps.check_step.outputs.branch }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Get current branch
+        id: check_step
+        run: |
+          if ${{ github.event_name == 'pull_request' }}
+          then
+            branch=${{ github.base_ref }}
+          else
+            raw=$(git branch -r --contains ${{ github.ref }})
+            branch=${raw##*/}
+          fi
+          echo "branch=$branch" >> $GITHUB_OUTPUT
+          echo "Branch is $branch."
+
+  build:
+    needs: check # Wait for check step to finish
+    runs-on: ubuntu-latest
+    # minimal permissions for Get Artifact Publish Token
+    permissions:
+      contents: read
+      id-token: write
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: 1.19.7
+      - name: Install Goveralls
+        run: go get -v github.com/mattn/goveralls
+      - name: Setup CI
+        run: make setup-ci
+      - name: Test
+        run: |
+          make check-fmt
+          make build-all
+          make check
+          make test-race
+          make bench-race
+      - name: Non Tag
+        if: github.ref_type != 'tag' || (needs.check.outputs.branch != 'master' && needs.check.outputs.branch != 'main' && needs.check.outputs.branch != 'release')
+        run: |
+          echo "Will not proceed to release job. Build does not meet requirement for release:"
+          echo "Ref type '$GITHUB_REF_TYPE' must equal tag and branch '${{needs.check.outputs.branch}}' must be master, main or release."
+
+  release:
+    needs:
+      - build
+      - check
+    if: github.ref_type == 'tag' && (needs.check.outputs.branch == 'master' || needs.check.outputs.branch == 'main' || needs.check.outputs.branch == 'release')
+    runs-on: ubuntu-latest
+    # minimal permissions for Get Artifact Publish Token
+    permissions:
+      contents: read
+      id-token: write
```

Comment:
@tiedotguy I'm looking into the way the Github does secrets, etc because you were asking about that before but the comment seems to have disappeared.

From my own usage of Github actions I know that you can set repository level secrets, but it seems like we don't have that kind of access, so let's assume we won't be using those.

These two permissions are for giving the Action the access it need to do its job...
`contents: read` are used to checkout the repo into the build environment
`id-token:write` are used to generate a JWT token, presumably unique to this repo.
See here: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#adding-permissions-settings

So, I'm not sure how travisci worked (will have to look at the config) but it seems like we might need to set up access to push the image to `Dockerhub.com/atlassianlabs/` via JWT auth. It might the the case that any JWT from this org is permitted?

You were also asking about fork access to secrets: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token
> Maximum access for pull requests from public forked repositories [1]

This is `read` for `id-token`, which paradoxically means they don't get access to it.
The [1] there means:
> Private repositories can control if pull requests from forks can run workflows, and configure the permissions assigned to GITHUB_TOKEN. For more information, see "[Managing GitHub Actions settings for a repository](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#enabling-workflows-for-forks-of-private-repositories)."

From the documentation is looks like you need to go in and explicitly allow fork pull requests permission to run actions, via the settings page.